products/cloudiot/api.yaml

```diff
@@ -192,3 +192,213 @@ objects:
                       output: true
                       description: |
                         The type of public key in the certificate.
+  - !ruby/object:Api::Resource
+    name: 'Device'
+    base_url: '{{registry}}/devices'
+    self_link: '{{registry}}/devices/{{name}}'
+    update_verb: :PATCH
+    update_mask: true
+    description: |
+      A Google Cloud IoT Core device.
+    references: !ruby/object:Api::Resource::ReferenceLinks
+      guides:
+        'Official Documentation':
+          'https://cloud.google.com/iot/docs/'
+      api: 'https://cloud.google.com/iot/docs/reference/cloudiot/rest/'
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: registry
+        input: true
+        url_param_only: true
+        required: true
+        description: |
+          The name of the device registry where this device should be created.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'id'
+        input: true
+        required: true
+        description: |
+          The unique identifier for the device. For example,
+          `Device0`.
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          The resource path name. For example,
+          `projects/example-proj/locations/us-central1/registries/my-registry/devices/device0`.
+      - !ruby/object:Api::Type::String
+        name: 'numId'
+        output: true
+        description: |
+          A server-defined unique numeric ID for the device.
+          This is a more compact way to identify devices, and it is globally unique.
+      - !ruby/object:Api::Type::Array
+        name: 'credentials'
+        description: |
+          The credentials used to authenticate this device.
+        max_size: 3
+        item_type: !ruby/object:Api::Type::NestedObject
+          properties:
+            - !ruby/object:Api::Type::Time
+              name: 'expirationTime'
+              allow_empty_object: true
+              output: true
+              description: |
+                The time at which this credential becomes invalid.
+            - !ruby/object:Api::Type::NestedObject
+              name: 'publicKey'
```

Comment:
This field should probably be `required`, or else a user can specify an empty `credentials` object which would be strange.

Reply:
The API and Cloud Console allow it. I think they have a point for a couple of use cases:

- Decoupling a device provisioning with credentials management.

- Supporting registry pre-provisioning with devices, where you don't authenticate devices yet. Example: when you manufacture a device, you create the entry in a registry, then when the user activates it, you upload the credentials.

I feel we shouldn't behave differently here.

Reply:
In terraform `required` on a nested object means that if that object is defined this field must be defined.

For example, in the current implementation it would be possible for a user to specify
```
credentials = {}
```

Which is not ideal. Users will still be able to not specify `credentials` even if this field is marked as required, we just want to make sure that they cannot specify an empty block

Reply:
ah! I get it now :)

Done